File: src/policy/predecodepolicy.rs
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

#[path = "predecodepolicy/cache_retention.rs"]
mod cache_retention;
#[path = "predecodepolicy/config.rs"]
mod config;
#[path = "predecodepolicy/load_queue.rs"]
mod load_queue;
#[path = "predecodepolicy/momentum.rs"]
mod momentum;
#[path = "predecodepolicy/schedule.rs"]
mod schedule;

#[cxx::bridge(namespace = "KiriView")]
mod ffi {
    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeSourceProfile {
        neutral_previous_image_count: usize,
        neutral_next_image_count: usize,
        biased_direction_image_count: usize,
        parallel_limit: usize,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustPredecodeMomentumMode {
        Neutral = 0,
        NextBiased = 1,
        PrevBiased = 2,
        ScrubbingNext = 3,
        ScrubbingPrev = 4,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    enum RustPredecodeMomentumDirection {
        None = 0,
        Previous = 1,
        Next = 2,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodePolicyInput {
        source_profile: RustPredecodeSourceProfile,
        momentum_mode: RustPredecodeMomentumMode,
        power_saver_enabled: bool,
    }

    #[derive(Debug, PartialEq, Eq)]
    struct RustPredecodeSchedulePlan {
        parallel_limit: usize,
        target_indices: Vec<usize>,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeMomentumState {
        last_page_index: i32,
        last_navigation_msec: i64,
        same_direction_move_count: i32,
        last_direction: RustPredecodeMomentumDirection,
        mode: RustPredecodeMomentumMode,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeMomentumUpdateInput {
        state: RustPredecodeMomentumState,
        page_index: i32,
        monotonic_msec: i64,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeQueuedLoadPlan {
        found: bool,
        index: usize,
        discard_count: usize,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeCachedImageState {
        current_displayed: bool,
        recent_displayed: bool,
        current_displayed_priority: usize,
        recent_displayed_priority: usize,
        window_priority: usize,
        byte_cost: i64,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeWindowLoadState {
        displayed: bool,
        cached: bool,
        in_flight: bool,
    }

    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
    struct RustPredecodeQueuedLoadState {
        valid: bool,
        in_window: bool,
        cached: bool,
        in_flight: bool,
    }

    extern "Rust" {
        #[cxx_name = "rustPredecodeDebounceMsec"]
        fn rust_predecode_debounce_msec() -> i32;

        #[cxx_name = "rustPredecodeNeutralRefreshMsec"]
        fn rust_predecode_neutral_refresh_msec() -> i32;

        #[cxx_name = "rustPredecodeRetainedCachedImageIndices"]
        fn rust_predecode_retained_cached_image_indices(
            states: Vec<RustPredecodeCachedImageState>,
            window_count: usize,
            byte_budget: i64,
        ) -> Vec<usize>;

        #[cxx_name = "rustPredecodeMissingWindowLoadIndices"]
        fn rust_predecode_missing_window_load_indices(
            states: Vec<RustPredecodeWindowLoadState>,
        ) -> Vec<usize>;

        #[cxx_name = "rustPredecodeSchedulePlan"]
        fn rust_predecode_schedule_plan(
            candidate_count: usize,
            current_index_found: bool,
            current_index: usize,
            input: RustPredecodePolicyInput,
        ) -> RustPredecodeSchedulePlan;

        #[cxx_name = "rustPredecodeUpdatedMomentumState"]
        fn rust_predecode_updated_momentum_state(
            input: RustPredecodeMomentumUpdateInput,
        ) -> RustPredecodeMomentumState;

        #[cxx_name = "rustPredecodeNextQueuedLoadPlan"]
        fn rust_predecode_next_queued_load_plan(
            states: Vec<RustPredecodeQueuedLoadState>,
        ) -> RustPredecodeQueuedLoadPlan;
    }
}

pub(super) use ffi::{
    RustPredecodeCachedImageState, RustPredecodeMomentumDirection, RustPredecodeMomentumMode,
    RustPredecodeMomentumState, RustPredecodeMomentumUpdateInput, RustPredecodePolicyInput,
    RustPredecodeQueuedLoadPlan, RustPredecodeQueuedLoadState, RustPredecodeSchedulePlan,
    RustPredecodeSourceProfile, RustPredecodeWindowLoadState,
};

fn rust_predecode_debounce_msec() -> i32 {
    config::debounce_msec()
}

fn rust_predecode_neutral_refresh_msec() -> i32 {
    config::neutral_refresh_msec()
}

fn rust_predecode_retained_cached_image_indices(
    states: Vec<RustPredecodeCachedImageState>,
    window_count: usize,
    byte_budget: i64,
) -> Vec<usize> {
    cache_retention::retained_cached_image_indices(states, window_count, byte_budget)
}

fn rust_predecode_missing_window_load_indices(
    states: Vec<RustPredecodeWindowLoadState>,
) -> Vec<usize> {
    load_queue::missing_window_load_indices(states)
}

fn rust_predecode_schedule_plan(
    candidate_count: usize,
    current_index_found: bool,
    current_index: usize,
    input: RustPredecodePolicyInput,
) -> RustPredecodeSchedulePlan {
    schedule::schedule_plan(candidate_count, current_index_found, current_index, input)
}

fn rust_predecode_updated_momentum_state(
    input: RustPredecodeMomentumUpdateInput,
) -> RustPredecodeMomentumState {
    momentum::updated_momentum_state(input)
}

fn rust_predecode_next_queued_load_plan(
    states: Vec<RustPredecodeQueuedLoadState>,
) -> RustPredecodeQueuedLoadPlan {
    load_queue::next_queued_load_plan(states)
}

File: app/src/predecode/predecodepolicy.cpp
// SPDX-FileCopyrightText: 2026 KIM Hyunjae
// SPDX-License-Identifier: AGPL-3.0-or-later

#include "predecodepolicy.h"

#include <algorithm>
#include <cmath>
#include <limits>
#include <tuple>

namespace {
constexpr qint64 biasedNavigationMsec = 450;
constexpr qint64 scrubbingNavigationMsec = 120;
constexpr qint64 neutralNavigationMsec = 800;
constexpr int scrubbingPageJump = 3;

qsizetype saturatedAdd(qsizetype left, qsizetype right)
{
    return right > 0 && left > std::numeric_limits<qsizetype>::max() - right
        ? std::numeric_limits<qsizetype>::max()
        : left + right;
}

qint64 saturatedSubtract(qint64 left, qint64 right)
{
    if (right < 0 && left > std::numeric_limits<qint64>::max() + right) {
        return std::numeric_limits<qint64>::max();
    }
    if (right > 0 && left < std::numeric_limits<qint64>::min() + right) {
        return std::numeric_limits<qint64>::min();
    }
    return left - right;
}

int saturatedIncrement(int value)
{
    return value == std::numeric_limits<int>::max() ? value : value + 1;
}

struct RetainedImage
{
    std::size_t index = 0;
    kiriview::PredecodeCachedImageState state;
};

std::size_t retentionGroup(const RetainedImage& image, std::size_t windowCount)
{
    if (image.state.recentDisplayed) {
        return 0;
    }
    return image.state.windowPriority < windowCount ? 1 : 2;
}
}

namespace kiriview {
int predecodeDebounceMsec() { return static_cast<int>(scrubbingNavigationMsec); }

int predecodeNeutralRefreshMsec() { return static_cast<int>(neutralNavigationMsec); }

PredecodeSourceProfile predecodeSourceProfileForOpenedCollectionScope(
    const OpenedCollectionScopeLocation& openedCollectionScope, int idealThreadCount)
{
    if (openedCollectionScope.isEmpty()) {
        return directMediaPredecodeSourceProfile();
    }
    if (openedCollectionScope.isDirectory()) {
        return { 3, 4, 6, 2 };
    }
    return { 3, 4, 6, static_cast<std::size_t>(std::clamp(idealThreadCount / 2, 1, 4)) };
}

std::vector<std::size_t> predecodeRetainedCachedImageIndices(
    const std::vector<PredecodeCachedImageState>& states, std::size_t windowCount,
    qsizetype byteBudget)
{
    if (byteBudget <= 0) {
        return {};
    }
    std::vector<RetainedImage> current;
    std::vector<RetainedImage> other;
    for (std::size_t index = 0; index < states.size(); ++index) {
        if (states[index].byteCost <= 0) {
            continue;
        }
        (states[index].currentDisplayed ? current : other).push_back({ index, states[index] });
    }
    std::ranges::sort(current, [](const RetainedImage& left, const RetainedImage& right) {
        return std::tie(left.state.currentDisplayedPriority, left.index)
            < std::tie(right.state.currentDisplayedPriority, right.index);
    });
    qsizetype currentCost = 0;
    for (const RetainedImage& image : current) {
        currentCost = saturatedAdd(currentCost, image.state.byteCost);
    }
    const qsizetype adjacentBudget
        = std::max<qsizetype>(byteBudget - std::min(byteBudget, currentCost), 0);
    std::ranges::sort(other, [windowCount](const RetainedImage& left, const RetainedImage& right) {
        const std::size_t leftGroup = retentionGroup(left, windowCount);
        const std::size_t rightGroup = retentionGroup(right, windowCount);
        if (leftGroup != rightGroup) {
            return leftGroup < rightGroup;
        }
        if (leftGroup == 0
            && left.state.recentDisplayedPriority != right.state.recentDisplayedPriority) {
            return left.state.recentDisplayedPriority < right.state.recentDisplayedPriority;
        }
        if (leftGroup == 1 && left.state.windowPriority != right.state.windowPriority) {
            return left.state.windowPriority < right.state.windowPriority;
        }
        if (leftGroup == 2 && left.state.lastUsedSequence != right.state.lastUsedSequence) {
            return left.state.lastUsedSequence > right.state.lastUsedSequence;
        }
        return left.index < right.index;
    });
    qsizetype retainedCost = 0;
    for (const RetainedImage& image : other) {
        if (saturatedAdd(retainedCost, image.state.byteCost) > adjacentBudget) {
            break;
        }
        retainedCost = saturatedAdd(retainedCost, image.state.byteCost);
        current.push_back(image);
    }
    std::vector<std::size_t> indices;
    indices.reserve(current.size());
    for (const RetainedImage& image : current) {
        indices.push_back(image.index);
    }
    return indices;
}

std::vector<std::size_t> predecodeMissingWindowLoadIndices(
    const std::vector<PredecodeWindowLoadState>& states)
{
    std::vector<std::size_t> indices;
    for (std::size_t index = 0; index < states.size(); ++index) {
        if (!states[index].displayed && !states[index].cached && !states[index].inFlight) {
            indices.push_back(index);
        }
    }
    return indices;
}

PredecodeSchedulePlan predecodeSchedulePlan(
    std::size_t candidateCount, std::optional<std::size_t> currentIndex, PredecodePolicyInput input)
{
    if (input.powerSaverEnabled || input.momentumMode == PredecodeMomentumMode::ScrubbingNext
        || input.momentumMode == PredecodeMomentumMode::ScrubbingPrev) {
        return {};
    }
    PredecodeSchedulePlan plan { input.sourceProfile.parallelLimit, {} };
    if (plan.parallelLimit == 0 || !currentIndex.has_value() || *currentIndex >= candidateCount) {
        return plan;
    }
    plan.targetIndices.push_back(*currentIndex);
    std::size_t nextCount = input.sourceProfile.neutralNextPageCount;
    std::size_t previousCount = input.sourceProfile.neutralPreviousPageCount;
    if (input.momentumMode == PredecodeMomentumMode::NextBiased) {
        nextCount = input.sourceProfile.biasedDirectionPageCount;
        previousCount = 1;
    } else if (input.momentumMode == PredecodeMomentumMode::PrevBiased) {
        nextCount = 1;
        previousCount = input.sourceProfile.biasedDirectionPageCount;
    }
    for (std::size_t offset = 1; offset <= std::max(nextCount, previousCount); ++offset) {
        if (offset <= nextCount && offset <= std::numeric_limits<std::size_t>::max() - *currentIndex
            && *currentIndex + offset < candidateCount) {
            plan.targetIndices.push_back(*currentIndex + offset);
        }
        if (offset <= previousCount && offset <= *currentIndex) {
            plan.targetIndices.push_back(*currentIndex - offset);
        }
    }
    return plan;
}

PredecodeMomentumState predecodeUpdatedMomentumState(
    PredecodeMomentumState state, int pageIndex, qint64 monotonicMsec)
{
    state.mode = PredecodeMomentumMode::Neutral;
    if (pageIndex < 0) {
        return state;
    }
    if (state.lastPageIndex < 0 || state.lastNavigationMsec < 0) {
        state.lastPageIndex = pageIndex;
        state.lastNavigationMsec = monotonicMsec;
        return state;
    }
    if (pageIndex == state.lastPageIndex) {
        return state;
    }
    const qint64 elapsed = saturatedSubtract(monotonicMsec, state.lastNavigationMsec);
    const PredecodeMomentumDirection direction = pageIndex > state.lastPageIndex
        ? PredecodeMomentumDirection::Next
        : PredecodeMomentumDirection::Previous;
    state.sameDirectionMoveCount
        = direction == state.lastDirection && elapsed <= biasedNavigationMsec
        ? saturatedIncrement(state.sameDirectionMoveCount)
        : 1;
    const qint64 pageDelta = static_cast<qint64>(pageIndex) - state.lastPageIndex;
    if (elapsed >= 0 && elapsed < neutralNavigationMsec) {
        const bool next = pageDelta > 0;
        const bool scrubbing
            = (state.sameDirectionMoveCount >= 2 && elapsed <= scrubbingNavigationMsec)
            || (std::abs(pageDelta) >= scrubbingPageJump && elapsed <= biasedNavigationMsec);
        if (scrubbing) {
            state.mode = next ? PredecodeMomentumMode::ScrubbingNext
                              : PredecodeMomentumMode::ScrubbingPrev;
        } else if (state.sameDirectionMoveCount >= 2 && elapsed <= biasedNavigationMsec) {
            state.mode
                = next ? PredecodeMomentumMode::NextBiased : PredecodeMomentumMode::PrevBiased;
        }
    }
    state.lastDirection = direction;
    state.lastPageIndex = pageIndex;
    state.lastNavigationMsec = monotonicMsec;
    return state;
}

PredecodeQueuedLoadPlan predecodeNextQueuedLoadPlan(
    const std::vector<PredecodeQueuedLoadState>& states)
{
    for (std::size_t index = 0; index < states.size(); ++index) {
        const PredecodeQueuedLoadState& state = states[index];
        if (state.valid && state.inWindow && !state.cached && !state.inFlight) {
            return { true, index,
                index == std::numeric_limits<std::size_t>::max() ? index : index + 1 };
        }
    }
    return { false, 0, states.size() };
}
}
